lib/src/utils.dart

```diff
@@ -82,3 +83,73 @@ void logDependencyInfractions(String infraction, Iterable<String> dependencies)
   final sortedDependencies = dependencies.toList()..sort();
   logger.warning([infraction, bulletItems(sortedDependencies), ''].join('\n'));
 }
+
+/// Lists the packages with infractions
+List<String> getDependenciesWithPins(Map dependencies) {
+  final List<String> infractions = [];
+  for (String packageName in dependencies.keys) {
+    String version;
+    if (dependencies[packageName] is String) {
+      version = dependencies[packageName];
+    } else {
+      final Map packageMeta = dependencies[packageName];
+      if (packageMeta.containsKey('version')) {
+        version = packageMeta['version'];
+      } else {
+        // This feature only works for versions, not git refs or paths.
+        break;
+      }
+    }
+
+    if (doesVersionPinDependency(version)) {
+      infractions.add(packageName);
+    }
+  }
+
+  return infractions;
+}
+
+/// Returns whether the version restricts patch or minor upgrades.
+bool doesVersionPinDependency(String rawVersion) {
+  final String version = rawVersion.replaceAll('"', '').replaceAll("'", '');
+
+  final VersionConstraint constraint = new VersionConstraint.parse(version);
+
+  // No max set = no pin
+  if (constraint.isAny) {
+    return false;
+  }
+
+  // Direct pin
+  if (constraint is Version) {
+    return true;
+  }
+
+  if (constraint is VersionRange) {
+    // Defined maximum bound is a pin
+    if (constraint.includeMax) {
+      return true;
+    }
+
+    final Version max = constraint.max;
+
+    // Builds and prereleases also block patch and minor bumps
+    if (max.build.isNotEmpty ||
+        // First preRelease here is okay because the upper bound is not inclusive
+        (max.isPreRelease && !max.isFirstPreRelease)) {
+      return true;
+    }
+
+    // Check legal upper bound doesn't block minor bumps
+    if (max.major == 0) {
+      if (max.minor == 0) {
+        return true; // 0.0.x
+      }
+      return max.patch != 0; // 0.x.0
+    }
+    return max.minor + max.patch > 0; // x.0.0
```

Comment:
Is this just a tricky way to see whether either is non-zero? If so, I think the code would read more clearly if it was straightforward:
```dart
return max.minor != 0 || max.patch != 0;
```
 If not, what's going on?

Reply:
Yeah it was just code golfing. I can change it to more readable.

Reply:
Thanks! I appreciate the code golfing spirit, though 😄